background/index.js

```diff
@@ -17,13 +17,171 @@
 /* global chrome */
 
 import Processor from './processor';
+import Ws from './ws';
 
-const processor = new Processor();
+import { UI, TRANSPORT_UNINITIALIZED } from '../shared';
 
 chrome.runtime.onConnect.addListener((port) => {
-  console.assert(port.name === 'id');
+  if (port.name === 'secureApi') {
+    port.onMessage.addListener(web3Message(port));
+    return;
+  }
+
+  if (port.name === 'barScripts') {
+    port.onMessage.addListener(loadScripts(port));
+    return;
+  }
+
+  // TODO [ToDr] Use connection without elevated privileges here!
+  if (port.name === 'web3') {
+    port.onMessage.addListener(web3Message(port));
+    return;
+  }
+
+  if (port.name === 'id') {
+    port.onMessage.addListener(processId(port));
+    return;
+  }
+
+  throw new Error(`Unrecognized port: ${port.name}`);
+});
+
+let openedTabId = null;
+let transport = null;
+// Attempt to extract token on start if not available.
+extractToken();
+
+chrome.runtime.onMessage.addListener((request, sender, callback) => {
+  if (!(transport && transport.isConnected) && request.token) {
+    if (transport) {
+      // TODO [ToDr] kill old transport!
+    }
+
+    if (openedTabId) {
+      chrome.tabs.remove(openedTabId);
```

Comment:
Doesn't it looks a bit sketchy to randomly open and close the tab when the Extension installs ? The user won't have the time to understand what happens ; I know I would be surprised. Might be better to have a settings/options/status window where it says that it needs the auth token, and on a button click it could open the right tab and gather the right data. However this might be included in a future issue.